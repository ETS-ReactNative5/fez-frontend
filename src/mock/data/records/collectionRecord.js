import { hydrateMock } from '../../hydrateMock';

const collectionRecord = {
    rek_pid: 'UQ:11398',
    rek_title: 'This is a Collection Record',
    rek_display_type: 9,
    rek_display_type_lookup: 'Collection',
    rek_status: 2,
    rek_status_lookup: 'Published',
    rek_object_type: 2,
    rek_object_type_lookup: 'Collection',
    fez_record_search_key_xsd_display_option: [263],
    rek_wok_doc_type_lookup: 'Abstract of Published Item',
};
export default hydrateMock(collectionRecord);
