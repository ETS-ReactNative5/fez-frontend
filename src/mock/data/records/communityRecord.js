import { hydrateMock } from '../../hydrateMock';

const communityRecord = {
    "rek_pid": "UQ:3883",
    "rek_title": "The University of Queensland Library",
    "rek_description": null,
    "rek_display_type": 11,
    "rek_status": 2,
    "rek_date": null,
    "rek_object_type": 1,
    "rek_security_inherited": 0,
    "rek_security_policy": 5,
    "rek_display_type_lookup": "Community",
    "rek_object_type_lookup": "Community",
    "rek_status_lookup": "Published",
};
export default hydrateMock(communityRecord).data;
