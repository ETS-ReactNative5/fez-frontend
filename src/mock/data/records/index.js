export { default as publicationTypeListAudio } from './publicationTypeListAudio';
export { default as publicationTypeListBook } from './publicationTypeListBook';
export { default as publicationTypeListBookChapter } from './publicationTypeListBookChapter';
export { default as publicationTypeListConferencePaper } from './publicationTypeListConferencePaper';
export { default as publicationTypeListConferenceProceedings } from './publicationTypeListConferenceProceedings';
export { default as publicationTypeListCreativeWork } from './publicationTypeListCreativeWork';
export { default as publicationTypeListDepartmentTechnicalReport } from './publicationTypeListDepartmentTechnicalReport';
export { default as publicationTypeListImage } from './publicationTypeListImage';
export { default as publicationTypeListJournal } from './publicationTypeListJournal';
export { default as publicationTypeListManuscript } from './publicationTypeListManuscript';
export { default as publicationTypeListNewspaperArticle } from './publicationTypeListNewspaperArticle';
export { default as publicationTypeListPatent } from './publicationTypeListPatent';
export { default as publicationTypeListPreprint } from './publicationTypeListPreprint';
export { default as publicationTypeListResearchReport } from './publicationTypeListResearchReport';
export { default as publicationTypeListSeminarPaper } from './publicationTypeListSeminarPaper';
export { default as publicationTypeListThesis } from './publicationTypeListThesis';
export { default as publicationTypeListVideo } from './publicationTypeListVideo';
export { default as publicationTypeListWorkingPaper } from './publicationTypeListWorkingPaper';
export { default as collectionRecord } from './collectionRecord';
export { default as collections } from './collections';
export { default as collectionsByCommunity } from './collectionsByCommunity';
export { default as communityRecord } from './communityRecord';
export { default as communitySearchList } from './communitySearchList';
export { default as externalDoiSearchResultList } from './externalDoiSearchResultList';
export { default as externalPubMedSearchResultsList } from './externalPubMedSearchResultsList';
export { default as externalTitleScopusResultsList } from './externalTitleScopusResultsList';
export { default as externalTitleSearchResultsList } from './externalTitleSearchResultsList';
export { default as incompleteNTROlist } from './incompleteNTROlist';
export { default as incompleteNTROrecord } from './incompleteNTROrecord';
export { default as incompleteNTROrecordUqrdav10 } from './incompleteNTRORecordUqrdav10';
export { default as incompleteNTROrecordUqsbutl1 } from './incompleteNTRORecordUqsbutl1';
export { default as internalTitleSearchList } from './internalTitleSearchList';
export { default as internalTitleSearchListNoResults } from './internalTitleSearchListNoResults';
export { default as lookupToolIncites } from './lookupToolIncites';
export { default as MyDatasetList } from './MyDatasetList';
export { default as myRecordsList } from './myRecordsList';
export { default as possibleUnclaimedList } from './possibleUnclaimedList';
export { default as record } from './record';
export { default as recordsTypeList } from './recordsTypeList';
export { default as recordWithDatastreams } from './recordWithDatastreams';
export { default as recordWithoutAuthorIds } from './recordWithoutAuthorIds';
export { default as recordWithTiffAndThumbnail } from './recordWithTiffAndThumbnail';
export { default as UQ352045 } from './incompleteNTRORecordUQ352045';
export { default as UQ716942uqagrinb } from './incompleteNTRORecordUQ716942_uqagrinb';
export { default as UQ716942uqagrinbGrants } from './incompleteNTRORecordUQ716942_uqagrinb_grants';
