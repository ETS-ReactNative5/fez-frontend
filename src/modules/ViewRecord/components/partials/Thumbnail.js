import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import ExternalLink from 'modules/SharedComponents/ExternalLink/components/ExternalLink';
import {BrokenImage} from '@material-ui/icons';
import {withStyles} from '@material-ui/core';
import locale from 'locale/pages';

const styles = () => ({
    brokenImage: {
        opacity: 0.5,
    }
});

class Thumbnail extends PureComponent {
    static propTypes = {
        mediaUrl: PropTypes.string.isRequired,
        previewMediaUrl: PropTypes.string.isRequired,
        thumbnailMediaUrl: PropTypes.string.isRequired,
        thumbnailFileName: PropTypes.string.isRequired,
        fileName: PropTypes.string,
        mimeType: PropTypes.string.isRequired,
        onClick: PropTypes.func,
        classes: PropTypes.object
    };

    constructor(props) {
        super(props);
        this.state = {
            thumbnailError: false
        };
    }

    showPreview = (mediaUrl, previewMediaUrl, mimeType) => (e) => {
        e.preventDefault();
        this.props.onClick(mediaUrl, previewMediaUrl, mimeType);
    };

    imageError = () => {
        this.setState({
            thumbnailError: true
        });
    };

    render() {
        const txt = locale.pages.viewRecord;
        const {mediaUrl, thumbnailMediaUrl, thumbnailFileName, previewMediaUrl, fileName, mimeType} = this.props;

        // TODO revert once videos are transcoded to open format #158519502
        if (fileName && (mimeType.indexOf('video') >= 0 || mimeType.indexOf('octet-stream') >= 0)) {
            return (
                !this.state.thumbnailError ?
                    <ExternalLink href={mediaUrl} title={fileName}  openInNewIcon={false}>
                        <img src={thumbnailMediaUrl} alt={thumbnailFileName} onError={this.imageError}/>
                    </ExternalLink>
                    : <BrokenImage color={'secondary'} />
            );
        }

        return (
            <a
                onClick={this.showPreview(mediaUrl, previewMediaUrl, mimeType)}
                onKeyPress={this.showPreview(mediaUrl, previewMediaUrl, mimeType)}
                title={mediaUrl && txt.thumbnailTitle.replace('[image]', mediaUrl) || null}
            >
                {
                    !this.state.thumbnailError ?
                        <img src={thumbnailMediaUrl} alt={thumbnailFileName} onError={this.imageError}/>
                        : <BrokenImage color={'secondary'} className={this.props.classes.brokenImage}/>
                }
            </a>
        );
    }
}

export default withStyles(styles)(Thumbnail);
