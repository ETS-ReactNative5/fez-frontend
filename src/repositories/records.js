import {post, patch} from './generic';
import * as routes from './routes';

/**
 * Posts a records item
 * @param {object} data to be posted, refer to backend API
 * @returns {Promise}
 */
export function postRecord(data) {
    if (data.rek_pid) return Promise.resolve(data);
    return post(routes.NEW_RECORD_API(), data);
}

/**
 * Patches a records item
 * @param {string} pid of object to be updated
 * @param {object} data to be posted, refer to backend API
 * @returns {Promise}
 */
export function patchRecord(pid, data) {
    return patch(routes.EXISTING_RECORD_API({pid: pid}), data);
}
