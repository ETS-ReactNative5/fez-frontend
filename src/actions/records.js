import {post, patch} from 'repositories/generic';
import * as routes from 'repositories/routes';
import {putUploadFiles} from '../repositories';
import * as transformers from './transformers';
import {NEW_RECORD_DEFAULT_VALUES} from 'config/general';
import * as actions from './actionTypes';

/**
 * Save a new record involves up to three steps: create a new record, upload files, update record with uploaded files.
 * If error occurs on any stage failed action is dispatched
 * @param {object} data to be posted, refer to backend API
 * @returns {promise} - this method is used by redux form onSubmit which requires Promise resolve/reject as a return
 */
export function createNewRecord(data) {
    return dispatch => {
        dispatch({type: actions.CREATE_RECORD_SAVING});

        // set default values, links
        const recordRequest = {
            ...NEW_RECORD_DEFAULT_VALUES,
            ...JSON.parse(JSON.stringify(data)),
            ...transformers.getRecordLinkSearchKey(data),
            ...transformers.getRecordAuthorsSearchKey(data.authors || data.currentAuthor),
            ...transformers.getRecordAuthorsIdSearchKey(data.authors || data.currentAuthor),
            ...transformers.getRecordContributorsSearchKey(data.editors),
            ...transformers.getRecordContributorsIdSearchKey(data.editors),
            ...transformers.getRecordSupervisorsSearchKey(data.supervisors),
            ...transformers.getRecordSubjectSearchKey(data.fieldOfResearch)
        };

        // delete extra form values from request object
        if (recordRequest.authors) delete recordRequest.authors;
        if (recordRequest.editors) delete recordRequest.editors;
        if (recordRequest.files) delete recordRequest.files;
        if (recordRequest.currentAuthor) delete recordRequest.currentAuthor;
        if (recordRequest.supervisors) delete recordRequest.supervisors;
        if (recordRequest.fieldOfResearch) delete recordRequest.fieldOfResearch;

        let newRecord = null;
        const hasFilesToUpload = data.files && data.files.queue && data.files.queue.length > 0;
        const recordPatch = hasFilesToUpload ? {...transformers.getRecordFileAttachmentSearchKey(data.files.queue)} : null;

        return post(routes.NEW_RECORD_API(), recordRequest)
            .then(response => {
                // save new record response
                newRecord = response.data;
                return response;
            })
            .then(() =>(hasFilesToUpload ? putUploadFiles(newRecord.rek_pid, data.files.queue, dispatch) : newRecord))
            .then(() => (hasFilesToUpload ? patch(routes.EXISTING_RECORD_API({pid: newRecord.rek_pid}), recordPatch) : newRecord))
            .then((response) => {
                dispatch({
                    type: actions.CREATE_RECORD_SUCCESS,
                    payload: response.data ? response.data : newRecord
                });
                return Promise.resolve(response.data ? response.data : newRecord);
            })
            .catch(error => {
                // record was created, but file upload or record patch failed
                if (!!newRecord && !!newRecord.rek_pid) {
                    dispatch({
                        type: actions.CREATE_RECORD_SUCCESS,
                        payload: {
                            newRecord: newRecord,
                            fileUploadFailed: true
                        }
                    });

                    return Promise.resolve(newRecord);
                }

                dispatch({
                    type: actions.CREATE_RECORD_FAILED,
                    payload: error.message
                });

                return Promise.reject(error);
            });
    };
}


/**
 * Submit thesis involves two steps: upload files, create record with uploaded files.
 * If error occurs on any stage failed action is dispatched
 * @param {object} data to be posted, refer to backend API
 * @returns {promise} - this method is used by redux form onSubmit which requires Promise resolve/reject as a return
 */
export function submitThesis(data, author) {
    return dispatch => {
        const hasFilesToUpload = data.files && data.files.queue && data.files.queue.length > 0;
        if (!hasFilesToUpload) {
            // reject thesis submission, files are required
            return Promise.reject('Please attach files to proceed with thesis submission');
        }
        // set default values, links
        const recordRequest = {
            ...JSON.parse(JSON.stringify(data)),
            ...transformers.getRecordAuthorsSearchKey(data.currentAuthor),
            ...transformers.getRecordAuthorsIdSearchKey(data.currentAuthor),
            ...transformers.getRecordSupervisorsSearchKey(data.supervisors),
            ...transformers.getRecordSubjectSearchKey(data.fieldOfResearch),
            ...transformers.getRecordFileAttachmentSearchKey(data.files.queue)
        };

        // delete extra form values from request object
        if (recordRequest.authors) delete recordRequest.authors;
        if (recordRequest.editors) delete recordRequest.editors;
        if (recordRequest.files) delete recordRequest.files;
        if (recordRequest.currentAuthor) delete recordRequest.currentAuthor;
        if (recordRequest.supervisors) delete recordRequest.supervisors;
        if (recordRequest.fieldOfResearch) delete recordRequest.fieldOfResearch;

        let fileUploadSucceeded = false;
        dispatch({type: actions.CREATE_RECORD_SAVING});

        // debugger;
        return putUploadFiles(`UQ:${author.aut_student_username}`, data.files.queue, dispatch)
            .then((response) => {
                fileUploadSucceeded = !!response;
                return post(routes.NEW_RECORD_API(), recordRequest);
            })
            .then(response => {
                dispatch({
                    type: actions.CREATE_RECORD_SUCCESS,
                    payload: {
                        newRecord: response
                    }
                });
                return response;
            })
            .catch(error => {
                const specificError = !fileUploadSucceeded
                    ? 'File upload failed. Issue has been created to notify eSpace administrators. '
                    : 'Error occurred while saving record to eSpace. ';
                const compositeError = `${specificError} ${ error.message ? `(${error.message})` : '' }`;

                dispatch({
                    type: actions.CREATE_RECORD_FAILED,
                    payload: compositeError
                });

                return Promise.reject(compositeError);
            });
    };
}
/**
 * Clear new record
 * @returns {action}
 */
export function clearNewRecord() {
    return dispatch => {
        dispatch({
            type: actions.CREATE_RECORD_RESET
        });
    };
}
