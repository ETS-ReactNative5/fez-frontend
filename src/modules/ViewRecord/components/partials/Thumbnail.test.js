import {journalArticle} from 'mock/data/testing/records';
import Thumbnail from "./Thumbnail";

function setup(testProps, isShallow = true){
    const props = {
        mediaUrl: '/test/mediaUrl',
        previewMediaUrl: '/test/preview/mediaUrl',
        thumbnailMediaUrl: '/test/thumbnail/mediaUrl',
        thumbnailFileName: 'thumbnail.jpg',
        mimeType: 'image/jpeg',
        onClick: jest.fn(),
        ...testProps,
    };
    return getElement(Thumbnail, props, isShallow);
}

describe('Thumbnail Component ', () => {
    it('should render component and display thumbnail', () => {
        const wrapper = setup({});
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should run onClick function on click', () => {
        const onClick = jest.fn();
        const wrapper = setup({onClick: onClick}, false);
        const element = wrapper.find('Thumbnail a');
        expect(toJson(wrapper)).toMatchSnapshot();
        element.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should run onFileSelect function on key press', () => {
        const onClick = jest.fn();
        const wrapper = setup({onClick: onClick}, false);
        const element = wrapper.find('Thumbnail a');
        expect(toJson(wrapper)).toMatchSnapshot();
        element.simulate('keyPress');
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should show a broken thumbnail icon when the thumbnail wont load.', () => {
        const onClick = jest.fn();
        const wrapper = setup({});
        wrapper.instance().setState({thumbnailError: true});
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
