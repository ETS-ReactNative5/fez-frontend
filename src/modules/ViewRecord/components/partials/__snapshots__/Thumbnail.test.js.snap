// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail component should render <ExternalLink> or <BrokenImage> for specific mime types 1`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of /test/mediaUrl"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should render <ExternalLink> or <BrokenImage> for specific mime types 2`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of /test/mediaUrl"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should render <ExternalLink> or <BrokenImage> for specific mime types 3`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of /test/mediaUrl"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should render <ExternalLink> or <BrokenImage> for specific mime types 4`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of /test/mediaUrl"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should render FLV files as a thumbnail only for specific mime types 1`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of video.flv"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should render PDF files as a thumbnail only plus link to original 1`] = `
<WithStyles(ExternalLink)
  className=""
  classes={Object {}}
  href="pdf.jpg"
  openInNewIcon={false}
  rel="noopener noreferrer"
  target="_blank"
  title="original.pdf"
>
  <Img
    alt="thumbnail.jpg"
    className=""
    container={[Function]}
    crossOrigin="anonymous"
    decode={true}
    loader={
      <ForwardRef(WithStyles)
        size={15}
        thickness={1}
      />
    }
    src="/test/thumbnail/mediaUrl"
    unloader={
      <Memo(BrokenImageIcon)
        color="secondary"
      />
    }
  />
</WithStyles(ExternalLink)>
`;

exports[`Thumbnail component should show a broken thumbnail icon when the thumbnail wont load. 1`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title="Click to open a preview of /test/mediaUrl"
>
  <LockIcon
    color="secondary"
  />
</a>
`;

exports[`Thumbnail component should show a broken thumbnail icon when the thumbnail wont load. 2`] = `
<a
  onClick={[Function]}
  onKeyPress={[Function]}
  title=""
>
  <LockIcon
    color="secondary"
  />
</a>
`;
